src/app/cli/src/coda_restarts_and_txns_holy_grail.ml

```diff
@@ -0,0 +1,74 @@
+open Core
+open Async
+open Coda_worker
+open Coda_main
+open Coda_base
+
+let name = "coda-restarts-and-txns-holy-grail"
+
+let main n () =
+  assert (n > 1) ;
+  let log = Logger.create () in
+  let log = Logger.child log name in
+  let snark_work_public_keys = function
+    | 0 ->
+        Some
+          (List.nth_exn Genesis_ledger.accounts 5 |> snd |> Account.public_key)
+    | _ -> None
+  in
+  let%bind testnet =
+    Coda_worker_testnet.test log n Option.some snark_work_public_keys
+      Protocols.Coda_pow.Work_selection.Seq
+  in
+  (* SEND TXNS *)
+  let sender_sk =
+    Genesis_ledger.largest_account_exn () |> fst |> Option.value_exn
+  in
+  let receiver_pk =
+    Genesis_ledger.largest_account_exn () |> snd |> Account.public_key
+  in
+  don't_wait_for
+  @@ Coda_worker_testnet.Payments.send_several_payments testnet ~node:0
+       ~src:sender_sk ~dest:receiver_pk ;
+  (* RESTART NODES *)
+  (* catchup *)
+  let idx = Quickcheck.random_value (Int.gen_incl 1 (n - 1)) in
+  let%bind () =
+    Coda_worker_testnet.Restarts.trigger_catchup testnet
+      ~largest_account_keypair:(Genesis_ledger.largest_account_keypair_exn ())
+      ~log ~node:idx
+      ~payment_receiver:(n - 1 - idx)
+  in
+  (* bootstrap *)
+  let idx = Quickcheck.random_value (Int.gen_incl 1 (n - 1)) in
+  let%bind () =
+    Coda_worker_testnet.Restarts.trigger_bootstrap testnet
+      ~largest_account_keypair:(Genesis_ledger.largest_account_keypair_exn ())
+      ~log ~node:idx
+      ~payment_receiver:(n - 1 - idx)
+  in
+  (* random *)
+  let idx = Quickcheck.random_value (Int.gen_incl 1 (n - 1)) in
+  let duration =
+    Quickcheck.(
+      random_value
+        Generator.(Float.gen_incl 1. 5. >>| fun x -> Time.Span.of_min x))
+  in
+  let%bind () =
+    Coda_worker_testnet.Restarts.restart_node testnet ~log ~node:idx
+      ~action:(Fn.const Deferred.unit) ~duration
+  in
+  (* settle for a few more min *)
+  after (Time.Span.of_min 3.)
```

Comment:
where do you check that catchup etc actually worked?

Reply:
In the current integration testing framework, all the checks happen via polling from within `Coda_worker_testnet.test`. These checks will be adjusted as we refine the other catchup-based tests and fix bugs. Most likely, we'll add some hooks at the beginning and end of the `restart_node` function to turn on and off those pieces of logic (I believe this is what @ghost-not-in-the-shell is looking into for the basic catchup test.

Reply:
can you add a comment saying (* TODO: Check catchup worked *) or something?